### PR TITLE
Added Formik

### PR DESCRIPTION
**Pull request by noranda**
NOTE: This branch is based off the `demographics` branch and should be merged after.

This PR adds [Formik](https://jaredpalmer.com/formik/), a widely-used React form library that 

> takes care of the repetitive and annoying stuff--keeping track of values/errors/visited fields, orchestrating validation, and handling submission--so you don't have to

@radamson and I discussed the possibility of using this library to handle tracking form fields and as our fields began to be more complex, it made sense to make the conversion before we got too much further. This will also help us in the future if we decide to add validation.

This PR specifically does the following:

- adds `formik`, `formik-material-ui`, and `formik-material-ui-pickers` to handle form data
- adds `downloadjs` to handle downloading the form data on submit
- adds a "Submit" button at the bottom of the form that will download the input form data as a JSON file
- converts all components to use Formik for form data tracking
- adds `AutocompleteField` (for selects with many options) and `TelField` form elements
- updates `DateField` to use a format placeholder
- auto inputs age and age units once a date of birth is entered
- uses `useMemo` react hooks where we want to limit rerunning of functions (ex. getting states/counties)
- uses react dom `batch` to do things like reset a field if another field is selected (ex. testing status)
- updates various dependencies
- updates styles and tests

### CODE REVIEW

**Comment by noranda** [comment]
There is an issue with IE11 not showing the placeholders on the `DateField` inputs. It's described [here](https://github.com/mui-org/material-ui-pickers/issues/621) but doesn't look to be resolved. I think it's fine to leave this for now, but we may want to revisit the issue in the future.

File: src/components/FormProvider/FormProvider.jsx [hunk 1]
```diff
@@ -1,13 +1,9 @@
...
 
-const FormContext = createContext([]);
-
-const FormProvider = ({ children, value }) => {
-  const [form, setForm] = useState(value ?? {});
-  return <FormContext.Provider value={{ form, setForm }}>{children}</FormContext.Provider>;
-};
-
-const useForm = () => useContext(FormContext);
+function FormProvider({ children }) {
```
Comment (radamson) [inline src/components/FormProvider/FormProvider.jsx]: Is this really still a form provider?  It seems like its more a date utils provider if anything.

I could see it still being a FormProvider if we moved `<Formik />` and maybe `<Form />` here, which I think would make sense.  That could also help reduce our coupling to this (or any other) implementation.  If we do this some of the removed tests could still apply, even if they'll be testing the library more than anything else (e.g. tests around intial values or state updates).

Comment (radamson) [reply src/components/FormProvider/FormProvider.jsx]: If we wanted to go crazy on that front setting we could also set`useForm = useFormikContext` to contain the Formik dependency, but I don't think that's necessary at this point.

File: src/components/sections/DemographicsSection/DemographicsSection.js [hunk 2]
```diff
@@ -39,20 +41,46 @@ const sexOptions = [
...
+  const { values, setFieldValue } = useFormikContext();
+
+  const handleBirthdayChange = useCallback(
+    date => {
+      batch(() => {
+        setFieldValue('dateOfBirth', date);
+        if (!isNaN(date)) {
+          const now = new Date();
+          const dayDiff = differenceInDays(now, date);
+          const monthDiff = differenceInMonths(now, date);
```
Comment (radamson) [inline src/components/sections/DemographicsSection/DemographicsSection.js]: should there be a `const yearDiff = differenceInYears(now, date)` for consistency?

File: src/components/forms/DateField/DateField.js [hunk 3]
```diff
@@ -1,38 +1,33 @@
...
+import { Field } from 'formik';
 import { KeyboardDatePicker } from '@material-ui/pickers';
+import { fieldToKeyboardDatePicker } from 'formik-material-ui-pickers';
+import FormControl from 'components/forms/FormControl';
 
-import useStyles from './styles';
-import { useForm } from 'components/FormProvider';
+function FormikKeyboardDatePicker({ children, onChange, ...props }) {
+  const datePickerProps = fieldToKeyboardDatePicker(props);
+  if (onChange) datePickerProps.onChange = onChange;
```
Comment (radamson) [inline src/components/forms/DateField/DateField.js]: No change here, but I imagine other form components might evolve to need this if they require greater control at some point?

**Comment by cmoesel** [comment]
> One thing I did notice in testing is that changing the age units doesn't change the age. I think we'll either need to lock the unit or change the age when that's changed.

I don't know...  I'm thinking that most of the time if I'm changing the unit it's because I selected the wrong one by mistake -- so it would be pretty annoying if I had put `60` `days` by mistake and when I changed it to `years` the `60` changed to `0`.  I think that might be too smart for our own good.

**Comment by cmoesel** [comment]
Also, @noranda -- you are a machine!

File: src/components/FormContent/FormContent.jsx [hunk 4]
```diff
@@ -24,22 +25,68 @@ const sections = [
...
+  CDC2019nCoVID: '',
+  NNDSSCaseId: '',
+  interviewerFirstName: '',
+  interviewerLastName: '',
+  interviewerAffiliationOrganization: '',
+  interviewerTelephone: '',
+  interviewerEmail: '',
+  currentStatus: '',
+  testingStatus: '',
+  puiReportDate: null,
```
[the remark this answers is not in the document]
Comment (noranda) [reply src/components/FormContent/FormContent.jsx]: Right now it's just what types the different Material UI elements allow. In my next PR, I've moved these to their own file, changed all but dates to strings (dates have to be null), and added a comment at the top explaining this.

File: src/components/FormContent/FormContent.jsx [hunk 5]
```diff
@@ -24,22 +25,68 @@ const sections = [
...
-    <div className={styles.root} key={s.id}>
-      <FormSectionHeader id={s.id} key={s.id} title={s.title} />
-      {s.section}
-    </div>
-  ));
+  return (
+    <Formik
+      initialValues={initialValues}
+      onSubmit={(values, actions) => {
+        download(JSON.stringify(values, 2, null), 'covid-19-pui-form.json', 'application/json');
```
[the remark this answers is not in the document]
Comment (noranda) [reply src/components/FormContent/FormContent.jsx]: You are correct! I will fix this.

File: src/components/FormProvider/FormProvider.jsx (hunk 1 above)
Comment (noranda) [reply src/components/FormProvider/FormProvider.jsx]: I thought about moving the date utils provider to the app and just getting rid of this file, but it's used so much in the tests that it seemed simpler to leave it be. I do think it makes sense to move the Formik-related providers here as well, so I'll do that.

File: src/components/sections/DemographicsSection/DemographicsSection.js (hunk 2 above)
Comment (noranda) [reply src/components/sections/DemographicsSection/DemographicsSection.js]: Yep! Will fix. :)

**Comment by radamson** [comment]
> > One thing I did notice in testing is that changing the age units doesn't change the age. I think we'll either need to lock the unit or change the age when that's changed.
> 
> I don't know... I'm thinking that most of the time if I'm changing the unit it's because I selected the wrong one by mistake -- so it would be pretty annoying if I had put `60` `days` by mistake and when I changed it to `years` the `60` changed to `0`. I think that might be too smart for our own good.

I agree that would be annoying.  I was more thinking about when someone gives you an actual date of birth and then the age is calculated.  e.g. its a 13 month old child and you enter 02/20/2019.  The app says 1 years old, but you want months.  if you switch it to months it will say 1 month.

Essentially the age match the DOB (if given), so maybe the right answer is that you shouldn't be able to put in a numerical age if a birthday is given.

File: src/components/forms/DateField/DateField.js (hunk 3 above)
Comment (noranda) [reply src/components/forms/DateField/DateField.js]: Yes! I've had to implement this in more fields in my next PR. Stay tuned. ;)

**Comment by noranda** [comment]
@radamson I'm not sure what you mean by:

> One thing I did notice in testing is that changing the age units doesn't change the age

Why should changing the age units change the age? If you are thinking we should convert the age when the user selects a different age unit, I think that's probably unwanted behavior.

Changing the date of birth will change both the age and age units to match, but I allow them to edit those fields after (although I'm not sure why they would). I could disable those fields after they're calculated, but that doesn't feel like we should enforce that behavior. Thoughts?

I will address the rest of your comments. Thanks!

**Comment by noranda** [comment]
@radamson 

> I was more thinking about when someone gives you an actual date of birth and then the age is calculated. e.g. its a 13 month old child and you enter 02/20/2019. The app says 1 years old, but you want months. if you switch it to months it will say 1 month.

Ah I see what you're saying. This is just a matter of where we want to put the cutoffs for age units. I still think it's ok as is, even if the user has to input the age manually. Considering the demographics for this form, I also think it's probably a super edge case we can safely ignore for now. ;)

**Comment by radamson** [comment]
That works for me.  My main concern is just that this allows users to input data that is clearly conflicting, but agree that is an edge case.  Even for more mainstream scenarios I think it might make sense to prevent them entering age if a DOB is provided, similar to other conditional components we have.

But I thats a discussion we can have another day.  Good work.

**Comment by cmoesel** [comment]
>Good work.

I second that!

**Pull request merged**